File: src/repotask/commands/skills.py
"""`repo-task skills` — generate the agent-facing entry points."""

from __future__ import annotations

from typing import Any

import typer
from rich.console import RenderableType
from rich.table import Table

from repotask.config import load_config
from repotask.output import emit, guard
from repotask.skills import generator

app = typer.Typer(no_args_is_help=True)


@app.command()
@guard
def sync(
    dry_run: bool = typer.Option(False, "--dry-run", help="Preview without writing."),
) -> None:
    """Write Claude Code skills and the AGENTS.md block for this project."""
    config = load_config()
    results = generator.sync(config, dry_run=dry_run)
    others = generator.find_agent_files(config.root)

    data = {
        "dryRun": dry_run,
        "files": [{"path": item.path, "action": item.action} for item in results],
        "otherAgentFiles": others,
        "note": (
            "Skills reference commands only — no project knowledge is copied into them, "
            "so they stay correct as the knowledge base changes."
        ),
    }
    emit("skills.sync", data, _render)


def _render(data: dict[str, Any]) -> RenderableType:
    table = Table(title="Generated agent skills" + (" (dry run)" if data["dryRun"] else ""))
    table.add_column("file", overflow="fold")
    table.add_column("action")
    for item in data["files"]:
        table.add_row(item["path"], item["action"])
    if data["otherAgentFiles"]:
        table.caption = (
            "Other agent instruction files found: "
            + ", ".join(data["otherAgentFiles"])
            + ". Point them at AGENTS.md."
        )
    return table

File: src/commands/skills.rs
//! `repo-task skills` — generate the agent-facing entry points.

use anyhow::Result;
use serde_json::{json, Value};

use crate::config;
use crate::output;
use crate::skills;

pub fn sync(dry_run: bool) -> Result<()> {
    let config = config::load()?;
    let results = skills::sync(&config, dry_run)?;
    let others = skills::find_agent_files(&config.root);

    let data = json!({
        "dryRun": dry_run,
        "files": results
            .iter()
            .map(|item| json!({"path": item.path, "action": item.action}))
            .collect::<Vec<_>>(),
        "otherAgentFiles": others,
        "note": "Skills reference commands only — no project knowledge is copied into them, \
                 so they stay correct as the knowledge base changes.",
    });
    output::emit("skills.sync", &data, |value| {
        let mut lines = vec![format!(
            "Generated agent skills{}",
            if value["dryRun"].as_bool().unwrap_or(false) {
                " (dry run)"
            } else {
                ""
            }
        )];
        if let Some(files) = value["files"].as_array() {
            for file in files {
                lines.push(format!(
                    "  {:<9} {}",
                    file["action"].as_str().unwrap_or(""),
                    file["path"].as_str().unwrap_or("")
                ));
            }
        }
        if let Some(others) = value["otherAgentFiles"].as_array() {
            if !others.is_empty() {
                let names: Vec<&str> = others.iter().filter_map(|item| item.as_str()).collect();
                lines.push(format!(
                    "\nOther agent instruction files found: {}. Point them at AGENTS.md.",
                    names.join(", ")
                ));
            }
        }
        lines.join("\n")
    });
    Ok(())
}
